_posts/2025-02-13-rfprs.md

```diff
@@ -0,0 +1,347 @@
+repo which I use for this site:
+
+![lazygit](/assets/images/lazygit.png)
+
+It's organized into different pages, each of which has a different responsibility:
+
+- changed files
+- branches and remotes
+- commit list
+- stash
+- changes diff
+
+Using this interface, I will frequently:
+
+- build commits for a RFPR after a `git reset` to the base branch.
+- add individual lines across multiple files into a new or existing commit.
+- reorder commits on a feature branch.
+- amend a commit with a particular change.
+- take a change out of a commit and create a new commit from it.
+- ... and much more
+
+This post does not intend to be a tutorial on how to use `lazygit`. You should start with the
+[Tutorials](https://github.com/jesseduffield/lazygit?tab=readme-ov-file#tutorials) to get a sense
+for it. Getting a hang of the keybinds takes a short while, and then you can configure it to your
+liking.
+
+Believe me when I say that using a tool like this gives you a lot of power. So much so that when I
+create a RFPR, I spend more time on writing the documentation for it than actually building the
+commits using `lazygit` -- it's that good.
+
+## About Commit Mistakes
+
+The above approach is really helpful in ensuring that reviewers do not spend time giving feedback
+for changes that will never merge. Imagine you have a PR with 10 changes. The 5th commit introduces
+a bug which is then fixed in the 8th commit. Assuming that the reviewer will be reviewing in commit
+order, they will likely spend time giving feedback about the bug in the 5th commit, only to become
+irritated that they spent that time unnecessarily.
+
+Backing out all of your changes and then building the commits before the PR is created avoids this
+problem entirely. Performing a `git reset` to `origin/main` will in effect merge your changes
+together, only leaving the correct, bug-free diff.
+
+If after you've created your commits you notice a bug in the 5th commit, you definitely do not need
+to start all over. You can fix the bug in the working tree and then in `lazygit` navigate to the
+commit where the bug is. Hitting `A` will amend the selected commit with your changes. There are a
+lot of nice-to-haves like this in `lazygit`.
+
+# Receiving Feedback
+
+Once you've posted the RFPR and need to make followup commits, those commits should also be RFPR
+commits. Other people will come back later and read through them and we should also respect their
+time.
+
+I would highly recommend *not* rebasing or rewriting commits after you start getting feedback. The
+reason for this is that reordering the commits or amending them, ends up creating new commits under
+the hood (even if `lazygit` made it super easy to do so). Because reviewers will leave feedback on
+each commit during the initial review, if you rewrite that commit as a followup and force push, the
+feedback they left becomes orphaned and out of context because that commit no longer exists.
```

Comment:
This is interesting. When I'm reviewing a PR that has good commits, I have another tab open with the PR's files tab open where I leave comments in-line. That's mostly because the way GitHub handles comments on a commit is really awkward and, frankly, I don't trust it. Maybe that's because I've been burned by comments I've left on commits that become orphaned and are lost.

Reply:
Man, we are all GitHub prisoners 😂 I think phabricator handles this particular scenario with more grace, but I only used it once, and that was in 2012 at Boundary. There was a lot at that time that it did *not* handle with grace, and I was relieved when GitHub PRs became the standard, even with its warts.